wire/encoder.go

```diff
@@ -0,0 +1,419 @@
+}
+
+func (e *Encoder) uint16(v uint16) (int, error) {
+	b := make([]byte, shortSize)
+	endian.PutUint16(b, v)
+	return e.buf.Write(b)
+}
+
+// Int32 encodes int32 value to voltdb wire protocol Integer. For a successful
+// encoding the number of bytes written is 4
+func (e *Encoder) Int32(v int32) (int, error) {
+	return e.uint32(uint32(v))
+}
+
+func (e *Encoder) uint32(v uint32) (int, error) {
+	b := make([]byte, integerSize)
+	endian.PutUint32(b, v)
+	return e.buf.Write(b)
+}
+
+// Int64 encodes int64 value into voltdb wire protocol Long. For a successful
+// encoding the number of bytes written is 8
+func (e *Encoder) Int64(v int64) (int, error) {
+	return e.uint64(uint64(v))
+}
+
+func (e *Encoder) uint64(v uint64) (int, error) {
+	b := make([]byte, longSize)
+	endian.PutUint64(b, v)
+	return e.buf.Write(b)
+}
+
+// Float64 encodes float64 value to voltdb wire protocol float type. This uses
+// math.Float64bits to covert v to uint64 which is encoded into []byte of size
+// 8.  For a successful encoding the number of bytes written is 8
+func (e *Encoder) Float64(v float64) (int, error) {
+	return e.uint64(math.Float64bits(v))
+}
+
+// Binary encodes []byte to voltdb wire protocol varbinary
+//
+// This first encodes the size of v as voltdb Short followed by raw bytes of v.
+func (e *Encoder) Binary(v []byte) (int, error) {
+	s, err := e.Int32(int32(len(v)))
+	if err != nil {
+		return 0, err
+	}
+	n, err := e.buf.Write(v)
+	if err != nil {
+		return 0, err
+	}
+	return s + n, nil
+}
+
+// Bool encodes bool values to voltdb wireprotocol boolean
+func (e *Encoder) Bool(v bool) (int, error) {
+	if v {
+		return e.Byte(0x1)
+	}
+	return e.Byte(0x0)
+}
+
+// String encodes strings to voltdb wire protocol string. A string is treated
+// like []byte. We first encode the size of the string, followed by the raw
+// bytes of the string.
+func (e *Encoder) String(v string) (int, error) {
+	return e.Binary([]byte(v))
+}
+
+// Time encodes time.Time value to voltdb wire protocol time.
+func (e *Encoder) Time(v time.Time) (int, error) {
+	nano := v.Round(time.Microsecond).UnixNano()
+	if v.IsZero() {
+		return e.Int64(math.MinInt64)
+	}
+	return e.Int64(nano / int64(time.Microsecond))
+}
+
+// Write implements io.Writer interface
+func (e *Encoder) Write(b []byte) (int, error) {
+	return e.buf.Write(b)
+}
+
+// Read implements io.Reader interface
+func (e *Encoder) Read(b []byte) (int, error) {
+	return e.buf.Read(b)
+}
+
+// Marshal encodes query arguments, there are values passed as driver.Value when
+// executing queries
+func (e *Encoder) Marshal(v interface{}) (int, error) {
+	switch x := v.(type) {
+	case bool:
+		return e.MarshalBool(x)
+	case int8:
+		return e.MarshalByte(x)
+	case int16:
+		return e.MarshalShort(x)
+	case int32:
+		return e.MarshalInt32(x)
+	case int64:
+		return e.MarshalInt64(x)
+	case float64:
+		return e.MarshalFloat64(x)
+	case string:
+		return e.MarshalString(x)
+	case time.Time:
+		return e.MarshalTime(x)
+	default:
```

Comment:
I took a look into the wire protocol again. The `Encoder.Marshal` method only encodes parameters, there is no mention of `Table` type as parameter, unless if I missed something.

My understanding is, we receive tables from the database we don't send them.